libnmstate/state.py

```diff
@@ -264,3 +269,41 @@ def dict_update(origin_data, to_merge_data):
         else:
             origin_data[key] = val
     return origin_data
+
+
+class LinuxBridgeState(object):
+    DEFAULT_STP_ENABLED = True
+    DEFAULT_PORT_STP_HAIRPIN_MODE = False
+    DEFAULT_PORT_STP_PATH_COST = 100
+    DEFAULT_PORT_STP_PRIORITY = 32
+
+    DEFAULT_PORT_STP_OPTIONS = {
+        LinuxBridge.PORT_STP_HAIRPIN_MODE: DEFAULT_PORT_STP_HAIRPIN_MODE,
+        LinuxBridge.PORT_STP_PATH_COST: DEFAULT_PORT_STP_PATH_COST,
+        LinuxBridge.PORT_STP_PRIORITY: DEFAULT_PORT_STP_PRIORITY,
+    }
+
+    STP_OPTIONS = [
+        LinuxBridge.STP_PRIORITY,
+        LinuxBridge.STP_FORWARD_DELAY,
+        LinuxBridge.STP_HELLO_TIME,
+        LinuxBridge.STP_MAX_AGE,
+    ]
+
+    @staticmethod
+    def canonicalize(iface_state):
+        """
+        This function is doing:
+            * Add default port options if missing.
+        """
+        if iface_state.get(InterfaceState.KEY) != InterfaceState.UP:
+            return
+        bridge_state = iface_state.get(LinuxBridge.CONFIG_SUBTREE, {})
+        stp_state = bridge_state.get(LinuxBridge.STP_SUBTREE, {})
+        stp_enabled = stp_state.get(
+            LinuxBridge.STP_ENABLED, LinuxBridgeState.DEFAULT_STP_ENABLED)
+        if stp_enabled:
+            for port_state in bridge_state.get(LinuxBridge.PORT_SUBTREE, []):
```

Comment:
I guess in my ideal world, the port subtree would a separate class/object that would handle canonicalization as I described above and Ports would be objects, too.